File: npm/examples/webhooks_e2e.ts
import { QuicknodeSdk, TemplateArgs, WebhookStartFrom } from "../sdk";
import type {
  CreateWebhookFromTemplateParams,
  UpdateWebhookParams,
} from "../sdk";

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

async function main() {
  const qn = QuicknodeSdk.fromEnv();

  const before = await qn.webhooks.listWebhooks();
  console.log(`webhooks before: ${before.data.length} (total=${before.pageInfo.total})`);

  const count = await qn.webhooks.getEnabledCount();
  console.log(`enabled count: ${count.total}`);

  const createParams: CreateWebhookFromTemplateParams = {
    name: "E2E Test Webhook",
    network: "ethereum-mainnet",
    destinationAttributes: {
      url: "https://webhook.site/ae19071a-2dcc-4035-9cdf-406dcb4719ef",
    },
    templateArgs: TemplateArgs.evmWalletFilter({
      wallets: ["0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48"],
    }),
  };

  const webhook = await qn.webhooks.createWebhookFromTemplate(createParams);
  const id = webhook.id;
  console.log(`created: ${id} | ${webhook.status}`);

  const fetched = await qn.webhooks.getWebhook(id);
  console.log(`fetched: ${fetched.id} | ${fetched.name}`);

  const updateParams: UpdateWebhookParams = {
    name: "E2E Test Webhook Updated",
  };
  const updated = await qn.webhooks.updateWebhook(id, updateParams);
  console.log(`updated name: ${updated.name}`);

  await qn.webhooks.pauseWebhook(id);
  console.log("paused");

  await qn.webhooks.activateWebhook(id, { startFrom: WebhookStartFrom.Latest });
  console.log("activated");

  await qn.webhooks.deleteWebhook(id);
  console.log(`deleted: ${id}`);
  await sleep(1000);

  // Exercise the evm-contract-events template, which carries the multi-word
  // eventHashes field. The API expects eventHashes on the wire.
  const ceCreateParams: CreateWebhookFromTemplateParams = {
    name: "E2E Test Webhook (evmContractEvents)",
    network: "ethereum-mainnet",
    destinationAttributes: {
      url: "https://webhook.site/ae19071a-2dcc-4035-9cdf-406dcb4719ef",
    },
    templateArgs: TemplateArgs.evmContractEvents({
      contracts: ["0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48"],
      eventHashes: [
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
      ],
    }),
  };
  const ceWebhook = await qn.webhooks.createWebhookFromTemplate(ceCreateParams);
  console.log(`created (evmContractEvents): ${ceWebhook.id} | ${ceWebhook.status}`);
  await qn.webhooks.deleteWebhook(ceWebhook.id);
  console.log(`deleted (evmContractEvents): ${ceWebhook.id}`);
  await sleep(1000);

  const after = await qn.webhooks.listWebhooks();
  console.log(`webhooks after: ${after.data.length} (total=${after.pageInfo.total})`);
}

main();

File: crates/core/examples/webhooks_e2e.rs
use std::{thread, time::Duration};

use quicknode_sdk::{
    webhooks::{
        ActivateWebhookParams, CreateWebhookFromTemplateParams, EvmWalletFilterTemplate,
        GetWebhooksParams, TemplateArgs, UpdateWebhookParams, WebhookDestinationAttributes,
        WebhookStartFrom,
    },
    QuickNodeSdk, SdkFullConfig,
};

#[tokio::main]
#[allow(clippy::unwrap_used, clippy::expect_used)]
async fn main() {
    let config = SdkFullConfig::from_env().expect("Config from env failed");
    let qn = QuickNodeSdk::new(&config).expect("sdk failed to initialize");

    let before = qn
        .webhooks
        .list_webhooks(&GetWebhooksParams::default())
        .await
        .expect("list_webhooks failed");
    println!("webhooks before: {}", before.data.len());

    let count = qn
        .webhooks
        .get_enabled_count()
        .await
        .expect("get_enabled_count failed");
    println!("enabled count: {}", count.total);

    let template_args = TemplateArgs::evm_wallet_filter(&EvmWalletFilterTemplate {
        wallets: vec!["0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48".to_string()],
    })
    .expect("template args are valid");

    let create_params = CreateWebhookFromTemplateParams {
        name: "E2E Test Webhook".to_string(),
        network: "ethereum-mainnet".to_string(),
        notification_email: None,
        destination_attributes: WebhookDestinationAttributes {
            url: "https://webhook.site/ae19071a-2dcc-4035-9cdf-406dcb4719ef".to_string(),
            security_token: None,
            compression: None,
        },
        template_args,
    };
    let webhook = qn
        .webhooks
        .create_webhook_from_template(&create_params)
        .await
        .expect("create_webhook_from_template failed");
    let id = webhook.id.clone();
    println!("created: {} | {}", id, webhook.status);
    thread::sleep(Duration::from_secs(1));

    let fetched = qn
        .webhooks
        .get_webhook(&id)
        .await
        .expect("get_webhook failed");
    println!("fetched: {} | {}", fetched.id, fetched.name);

    let update_params = UpdateWebhookParams {
        name: Some("E2E Test Webhook Updated".to_string()),
        ..Default::default()
    };
    let updated = qn
        .webhooks
        .update_webhook(&id, &update_params)
        .await
        .expect("update_webhook failed");
    println!("updated name: {}", updated.name);
    thread::sleep(Duration::from_secs(1));

    qn.webhooks
        .pause_webhook(&id)
        .await
        .expect("pause_webhook failed");
    println!("paused");

    let activate_params = ActivateWebhookParams {
        start_from: WebhookStartFrom::Latest,
    };
    qn.webhooks
        .activate_webhook(&id, &activate_params)
        .await
        .expect("activate_webhook failed");
    println!("activated");

    qn.webhooks
        .delete_webhook(&id)
        .await
        .expect("delete_webhook failed");
    println!("deleted: {id}");
    thread::sleep(Duration::from_secs(1));

    let after = qn
        .webhooks
        .list_webhooks(&GetWebhooksParams::default())
        .await
        .expect("list_webhooks failed");
    println!("webhooks after: {}", after.data.len());
}
